lib/tsd_client.rb

```diff
@@ -34,15 +38,26 @@ def query options
       end
     end
 
-    def put options
+    def put options, retries = @options[:retries]
       Timeout::timeout @options[:timeout] do
-        TCPSocket.open @options[:host], @options[:port] do |socket|
-          socket.puts Format.put options
-
-          response, _, _ = socket.recvmsg_nonblock rescue nil
+        begin
+          @socket.puts Format.put options
+          response, _, _ = @socket.recvmsg_nonblock rescue nil
           response.chomp if response
+        rescue Exception => e
+          if retries > 0
+            sleep @options[:retry_interval]
+            $stderr.puts "Error: #{e}, retrying \##{retries}"
+            retries -= 1
+            # try to reconnect
+            @socket = TCPSocket.open @options[:host], @options[:port]
```

Comment:
it seems like this should be a method, but ignore this if i'm being unreasonable =)